src/coreclr/jit/async.cpp

```diff
@@ -836,330 +1196,387 @@ void Async2Transformation::Transform(
+    }
 
-            LclVarDsc* dsc = m_comp->lvaGetDesc(inf.LclNum);
+    unsigned    resumeObjectArrLclNum = BAD_VAR_NUM;
+    BasicBlock* storeResultBB         = resumeBB;
 
-            GenTree* byteArr = m_comp->gtNewLclvNode(resumeByteArrLclNum, TYP_REF);
-            unsigned offset  = OFFSETOF__CORINFO_Array__data + inf.DataOffset;
-            GenTree* cns     = m_comp->gtNewIconNode((ssize_t)offset, TYP_I_IMPL);
-            GenTree* addr    = m_comp->gtNewOperNode(GT_ADD, TYP_BYREF, byteArr, cns);
+    if (layout.GCRefsCount > 0)
+    {
+        resumeObjectArrLclNum = GetGCDataArrayVar();
 
-            GenTree* value;
-            if ((dsc->TypeGet() == TYP_STRUCT) || dsc->IsImplicitByRef())
-            {
-                value = m_comp->gtNewBlkIndir(dsc->GetLayout(), addr, GTF_IND_NONFAULTING);
-            }
-            else
-            {
-                value = m_comp->gtNewIndir(dsc->TypeGet(), addr, GTF_IND_NONFAULTING);
-            }
+        GenTree* newContinuation = m_comp->gtNewLclvNode(m_comp->lvaAsyncContinuationArg, TYP_REF);
+        unsigned gcDataOffset    = m_comp->info.compCompHnd->getFieldOffset(m_async2Info.continuationGCDataFldHnd);
+        GenTree* gcDataInd       = LoadFromOffset(newContinuation, gcDataOffset, TYP_REF);
+        GenTree* storeAllocedObjectArr = m_comp->gtNewStoreLclVarNode(resumeObjectArrLclNum, gcDataInd);
+        LIR::AsRange(resumeBB).InsertAtEnd(LIR::SeqTree(m_comp, storeAllocedObjectArr));
 
-            GenTree* store;
-            if (dsc->IsImplicitByRef())
-            {
-                GenTree* baseAddr = m_comp->gtNewLclvNode(inf.LclNum, dsc->TypeGet());
-                // TODO-CQ: Incoming data has no non-null GC refs, so this does not need write barriers.
-                // Backend does not handle GTF_IND_TGT_NOT_HEAP for STORE_BLK.
-                store = m_comp->gtNewStoreBlkNode(dsc->GetLayout(), baseAddr, value,
-                                                  GTF_IND_NONFAULTING | GTF_IND_TGT_NOT_HEAP);
-            }
-            else
-            {
-                store = m_comp->gtNewStoreLclVarNode(inf.LclNum, value);
-            }
+        RestoreFromGCPointersOnResumption(resumeObjectArrLclNum, layout.Locals, resumeBB);
 
-            LIR::AsRange(resumeBB).InsertAtEnd(LIR::SeqTree(m_comp, store));
+        if (layout.ExceptionGCDataIndex != UINT_MAX)
+        {
+            storeResultBB = RethrowExceptionOnResumption(block, remainder, resumeObjectArrLclNum, layout, resumeBB);
         }
     }
 
-    unsigned    resumeObjectArrLclNum = BAD_VAR_NUM;
-    BasicBlock* storeResultBB         = resumeBB;
+    // Copy call return value.
+    if (layout.ReturnSize > 0)
+    {
+        CopyReturnValueOnResumption(call, callDefInfo, resumeByteArrLclNum, resumeObjectArrLclNum, layout,
+                                    storeResultBB);
+    }
 
-    if (gcRefsCount > 0)
+    return resumeBB;
```

Comment:
Perhaps worth assigning null to data array fields after deserializing them. The dispatcher will keep the current continuation alive until we return (since it needs `.Next`). That may take a while, but we no longer need the arrays, so no need to keep them alive.

Reply:
Since this is mostly refactoring, it would be ok to just add a comment or put this onto the "suggestions" pile.

Reply:
I think we can make it so that `DispatchContinuations` doesn't keep the strong reference to `continuation` beyond the resumption call.

Reply:
I added a TODO in `DispatchContinuations` about this.